Add GenericShardingSpec for generic tensor sharding.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58517 ShardedTensor framework for ChunkedShardingSpec
* **#57409 Add GenericShardingSpec for generic tensor sharding.**
* #55728 Introduce ChunkShardingSpec as a model sharding specification.

Full design: https://github.com/pytorch/pytorch/issues/55207

In https://github.com/pytorch/pytorch/issues/55207, we proposed
`MeshShardingSpec` as a generic sharding mechanism. However, that proposal does
not provide the flexibility to specify shards which have uneven
sizes/partitions and assumes even partitioning. Uneven partitioning is one of
the requirements of an internal use case.

As a result, instead of that we introduce a `GenericShardingSpec` which allows
specifying any arbitrary partitioning of a multi dimensional tensor. Basically
it specifies the start offsets of each shard and the length of each dim of the
shard allowing for greater flexibility

Differential Revision: [D28137616](https://our.internmc.facebook.com/intern/diff/D28137616/)

pritamdamania committed:
Add GenericShardingSpec for generic tensor sharding.

Full design: https://github.com/pytorch/pytorch/issues/55207

In https://github.com/pytorch/pytorch/issues/55207, we proposed
`MeshShardingSpec` as a generic sharding mechanism. However, that proposal does
not provide the flexibility to specify shards which have uneven
sizes/partitions and assumes even partitioning. Uneven partitioning is one of
the requirements of an internal use case.

As a result, instead of that we introduce a `GenericShardingSpec` which allows
specifying any arbitrary partitioning of a multi dimensional tensor. Basically
it specifies the start offsets of each shard and the length of each dim of the
shard allowing for greater flexibility

Differential Revision: [D28137616](https://our.internmc.facebook.com/intern/diff/D28137616/)

[ghstack-poisoned]

xush6528 commented on test/distributed/_sharding_spec/test_sharding_spec.py:
```diff
@@ -52,3 +53,164 @@ def test_chunked_sharding_spec(self):
             ChunkShardingSpec(0, ["rank:0/cuda:foo", "cuda:1"])
+
+    def test_generic_sharding_spec(self):
+        Shard = GenericShardingSpec.Shard
```
Why the Shard is defined as a nested class?
Do users need to use `Shard` this way?

pritamdamania87 replied:
I have refactored this further in https://github.com/pytorch/pytorch/pull/58517

pritamdamania committed:
Update on "Add GenericShardingSpec for generic tensor sharding."


Full design: https://github.com/pytorch/pytorch/issues/55207

In https://github.com/pytorch/pytorch/issues/55207, we proposed
`MeshShardingSpec` as a generic sharding mechanism. However, that proposal does
not provide the flexibility to specify shards which have uneven
sizes/partitions and assumes even partitioning. Uneven partitioning is one of
the requirements of an internal use case.

As a result, instead of that we introduce a `GenericShardingSpec` which allows
specifying any arbitrary partitioning of a multi dimensional tensor. Basically
it specifies the start offsets of each shard and the length of each dim of the
shard allowing for greater flexibility

Differential Revision: [D28137616](https://our.internmc.facebook.com/intern/diff/D28137616/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -96,3 +96,173 @@ def placement(self) -> ShardPlacement:
+            shard_offsets(List[int]): Offsets in the orignal tensor indicating
+                the start offsets for this shard. Should have the same rank as
+                the original tensor.
+            shard_lengths(List[int]): Lengths indicating the length of each
```
curious, do we have any requirements/guideline on length distribution across shards? I assume we encourage uniform distribution, but support arbitrary distribution?

pritamdamania87 replied:
There are no guidelines as such as part of the specification itself. I think when we build more higher level APIs on top of this, we could have guidelines at that level in terms of what is best for performance.

mrshenli commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -96,3 +96,173 @@ def placement(self) -> ShardPlacement:
         """
         return self._placement
+
+class GenericShardingSpec(PlacementSpec):
```
Will there be another MeshShardingSpec build on top of this, or do we expect users to configure GenericShardingSpec in a mesh manner if they need mesh?

pritamdamania87 replied:
My thought was to only have GenericShardingSpec since it covers the Mesh use case in a much more generic way. I feel it is also simpler to have just two specs, one which is simple (ChunkShardingSpec) and probably covers most use cases and then another one which is more generic and covers everything.

pritamdamania87 replied:
Thinking about this more I realized the following. MeshShardingSpec can be applied to any tensor (so can ChunkShardingSpec) but GenericShardingSpec can be only applied to specific tensors since it involves shard offsets and lengths. So you need a different spec for each tensor (in case of GenericShardingSpec) instead of having one spec (in case of MeshShardingSpec) for a bunch of tensors.

Although, I do feel we can probably start out with these two to begin with (Generic and Chunk) and then evaluate if we do indeed need another spec like Mesh.

SciPioneer replied:
I think `ChunkShardingSpec` and `MeshShardingSpec` have a few advantages over `GenericShardingSpec`, where the former defines a sharding function that can be applied to dynamic shapes, where the latter basically hardcodes a sharding function by enumerating the sharding outputs. 

1) If the shape of an input tensor cannot be predefined (dynamic tensor can be common in NLP), then  `MeshShardingSpec` cannot be replaced by `GenericShardingSpec`.
2) Given a list of devices, actually both `ChunkShardingSpec` and `MeshShardingSpec` can be describes as a generator that yields a stream of shards. This may help save some memory if we have a large number of shards.

pritamdamania87 replied:
@SciPioneer Agree with what you mentioned above, although we have a concrete use case for uneven/arbitrary shard sizes which MeshShardingSpec doesn't satisfy and we need something like GenericShardingSpec for this.

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -96,3 +96,173 @@ def placement(self) -> ShardPlacement:
+
+class GenericShardingSpec(PlacementSpec):
+
+    class Shard:
```
Since there can be a large number of `Shard` instance, better to add `__slots__` for memory optimization.

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -96,3 +96,173 @@ def placement(self) -> ShardPlacement:
+        self._shards = shards
+
+    @staticmethod
+    def _validate_non_overlapping(shards: List[Shard]):
```
Besides checking no overlapping, should we also check if the input tensor is entirely covered by all the shards? In other words, the union of all the shards should be equal to the input tensor right?

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -96,3 +96,173 @@ def placement(self) -> ShardPlacement:
+        for size in tensor_dims:
+            tensor_volume *= size
+
+        if total_shard_volume != tensor_volume:
```
If `total_shard_volum` < `tensor_volume`, better to help identify the gaps that are not covered by the shards. Otherwise it can be very hard for the developers to eyeball all the shards. Therefore, it can be valuable to create a separate util function for providing more informative error message.

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -96,3 +96,173 @@ def placement(self) -> ShardPlacement:
+        """
+        Ensures none of the shards overlap with each other.
+        """
+        for i in range(len(shards)):
```
This O(N^2) cost seems to be suboptimal. Should be faster if we sort the shards first. Probably add a TODO here to avoid missing the optimization opporunity?

pritamdamania committed:
Update on "Add GenericShardingSpec for generic tensor sharding."


Full design: https://github.com/pytorch/pytorch/issues/55207

In https://github.com/pytorch/pytorch/issues/55207, we proposed
`MeshShardingSpec` as a generic sharding mechanism. However, that proposal does
not provide the flexibility to specify shards which have uneven
sizes/partitions and assumes even partitioning. Uneven partitioning is one of
the requirements of an internal use case.

As a result, instead of that we introduce a `GenericShardingSpec` which allows
specifying any arbitrary partitioning of a multi dimensional tensor. Basically
it specifies the start offsets of each shard and the length of each dim of the
shard allowing for greater flexibility

Differential Revision: [D28137616](https://our.internmc.facebook.com/intern/diff/D28137616/)

[ghstack-poisoned]

SciPioneer commented:
It seems that you still need to fix or suppress a mypy lint error.

SciPioneer commented:
Personally I think `EnumerableShardingSpec` can be more descriptive than `GenericShardingSpec`, because this sharding basically enumerates all the shards, so the name implies how the sharding works. `Generic` is a bit too vague, and it may even give an impression that it's a base `ShardingSpec` class.

pritamdamania87 commented:
>Personally I think EnumerableShardingSpec can be more descriptive than GenericShardingSpec, because this sharding basically enumerates all the shards, so the name implies how the sharding works. Generic is a bit too vague, and it may even give an impression that it's a base ShardingSpec class.

`EnumerableShardingSpec` does sound like a better name than `GenericShardingSpec`, it was too generic ;)

pritamdamania committed:
Update on "Add GenericShardingSpec for generic tensor sharding."


Full design: https://github.com/pytorch/pytorch/issues/55207

In https://github.com/pytorch/pytorch/issues/55207, we proposed
`MeshShardingSpec` as a generic sharding mechanism. However, that proposal does
not provide the flexibility to specify shards which have uneven
sizes/partitions and assumes even partitioning. Uneven partitioning is one of
the requirements of an internal use case.

As a result, instead of that we introduce a `GenericShardingSpec` which allows
specifying any arbitrary partitioning of a multi dimensional tensor. Basically
it specifies the start offsets of each shard and the length of each dim of the
shard allowing for greater flexibility

Differential Revision: [D28137616](https://our.internmc.facebook.com/intern/diff/D28137616/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add GenericShardingSpec for generic tensor sharding."


Full design: https://github.com/pytorch/pytorch/issues/55207

In https://github.com/pytorch/pytorch/issues/55207, we proposed
`MeshShardingSpec` as a generic sharding mechanism. However, that proposal does
not provide the flexibility to specify shards which have uneven
sizes/partitions and assumes even partitioning. Uneven partitioning is one of
the requirements of an internal use case.

As a result, instead of that we introduce a `GenericShardingSpec` which allows
specifying any arbitrary partitioning of a multi dimensional tensor. Basically
it specifies the start offsets of each shard and the length of each dim of the
shard allowing for greater flexibility

Differential Revision: [D28137616](https://our.internmc.facebook.com/intern/diff/D28137616/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add GenericShardingSpec for generic tensor sharding."


Full design: https://github.com/pytorch/pytorch/issues/55207

In https://github.com/pytorch/pytorch/issues/55207, we proposed
`MeshShardingSpec` as a generic sharding mechanism. However, that proposal does
not provide the flexibility to specify shards which have uneven
sizes/partitions and assumes even partitioning. Uneven partitioning is one of
the requirements of an internal use case.

As a result, instead of that we introduce a `GenericShardingSpec` which allows
specifying any arbitrary partitioning of a multi dimensional tensor. Basically
it specifies the start offsets of each shard and the length of each dim of the
shard allowing for greater flexibility

Differential Revision: [D28137616](https://our.internmc.facebook.com/intern/diff/D28137616/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add GenericShardingSpec for generic tensor sharding."


Full design: https://github.com/pytorch/pytorch/issues/55207

In https://github.com/pytorch/pytorch/issues/55207, we proposed
`MeshShardingSpec` as a generic sharding mechanism. However, that proposal does
not provide the flexibility to specify shards which have uneven
sizes/partitions and assumes even partitioning. Uneven partitioning is one of
the requirements of an internal use case.

As a result, instead of that we introduce a `GenericShardingSpec` which allows
specifying any arbitrary partitioning of a multi dimensional tensor. Basically
it specifies the start offsets of each shard and the length of each dim of the
shard allowing for greater flexibility

Differential Revision: [D28137616](https://our.internmc.facebook.com/intern/diff/D28137616/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add GenericShardingSpec for generic tensor sharding."


Full design: https://github.com/pytorch/pytorch/issues/55207

In https://github.com/pytorch/pytorch/issues/55207, we proposed
`MeshShardingSpec` as a generic sharding mechanism. However, that proposal does
not provide the flexibility to specify shards which have uneven
sizes/partitions and assumes even partitioning. Uneven partitioning is one of
the requirements of an internal use case.

As a result, instead of that we introduce a `GenericShardingSpec` which allows
specifying any arbitrary partitioning of a multi dimensional tensor. Basically
it specifies the start offsets of each shard and the length of each dim of the
shard allowing for greater flexibility

Differential Revision: [D28137616](https://our.internmc.facebook.com/intern/diff/D28137616/)

[ghstack-poisoned]
